src/hidra/receiver/datareceiver.py

```diff
@@ -259,6 +260,97 @@ def stop(self):
         self.run_loop = False
 
 
+def run_plugin_thread(
+        plugin_name, plugin_config, target_dir, data_queue, log, event):
+    # Execute all plugin code on a separate thread to protect main thread from
+    # slow or blocking plugins
+    try:
+        plugin_m = import_module("plugins." + plugin_name)
+        plugin = plugin_m.Plugin(plugin_config)
+        plugin.setup()
+        log.info("Loading '%s' plugin", plugin_name)
+    except Exception:
+        log.error("Could not load '%s' plugin", plugin_name, exc_info=True)
+        return
+
+    while not event.is_set():
+        try:
+            data = data_queue.get(timeout=0.5)
+        except queue.Empty:
+            continue
+
+        try:
+            [metadata, data] = data
+            plugin.process(
+                local_path=generate_filepath(target_dir, metadata),
+                metadata=metadata,
+                data=data
+            )
+        except Exception:
+            log.error("Processing data with '%s' plugin failed.",
+                      plugin_name, exc_info=True)
+
+        data_queue.task_done()
+
+    try:
+        plugin.stop()
+    except Exception:
+        log.error(
+            "Error while stopping '%s' plugin", plugin_name, exc_info=True)
+
```

Comment:
I can't see or find the threaded execution of the plugin code here!?

Reply:
I tried to reword the description. Is this more clear now?